mail/src/test/java/com/sun/mail/util/logging/MailHandlerTest.java

```diff
@@ -475,6 +477,96 @@ private void testLoggable(Level lvl, LogRecord record) {
         assertEquals(em.exceptions.isEmpty(), true);
     }
 
+    @Test
+    public void testDeclaredClasses() throws Exception {
+    	//MailHandler init will pin all classes that it may use.
+    	//Check that classes are pinned in static fields or
+    	//that there is a field that is designed to hold a declared type.
+    	//This includes public properties of the Handler class too.
+        for (Class<?> k : MailHandler.class.getDeclaredClasses()) {
+    		boolean found = false;
+
+    		//Has a getter method matching type.
+        	for (Method m : Handler.class.getDeclaredMethods()) {
+        		if (!Modifier.isPublic(m.getModifiers())) {
+        			continue;
+        		}
+
+        		if (m.getReturnType().isAssignableFrom(k)) {
+        			found = true;
+        			break;
+        		}
+        	}
+
+        	if (found) {
+        		continue;
+        	}
+
+    		//Has a field that holds the type.
+        	for (Field f : MailHandler.class.getDeclaredFields()) {
+        		Class<?> t = f.getType();
+        		if (t == Object.class) {
+        			continue;
+        		}
+
+        		if (t.isAssignableFrom(k)) {
+        			//Check for static field holding declared class.
+        			if (Modifier.isStatic(f.getModifiers())
+        					&& Modifier.isFinal(f.getModifiers())) {
+        				f.setAccessible(true);
+        				Object o = f.get((Object) null);
```

Comment:
I'm thinking about starting using `MethodHandles` in tests to avoid possible access issues going forward, especially when one needs to call `setAccessible`. Something like: `MethodHandles.privateLookupIn(MailHandler.class, MethodHandles.lookup()).unreflectVarHandle(f).get()` should be valid replacement but that would tight tests to JDK 9+, so not sure that is good idea. What do you think?

Reply:
Is the minimum version for 1.x JDK7?  I was thinking of backporting this to 1.x and would like to avoid diverging the code at this time if 1x requires a min version less than JDK9.  There are a handful of tests where I call setAccessible(true) I would prefer to tackle them in a follow on PR.  That would also give me time to test with your suggested code.

The alternatives tests to accomplish the intended goal would be:
1. The only other way I can think to avoid  create a tests that [launches a new VM](https://stackoverflow.com/questions/23686948/loading-junit-tests-to-different-jvms) with verbose class loading turned on via MBeans.  Then scrape the output to ensure all the declared classes are loaded.
2. Relax the static field to default access (package protected) and inspect it.  I rather it be private but this is possible.  There are few more places that need promoted access if I go down this road.

So if you are good with it lets create a follow on PR to tackle use of setAccessible(true);

Reply:
> Is the minimum version for 1.x JDK7?

Yes, good point and should stay there forever... multi-release jars can be an option but I have no idea how they work in/for tests as typically one runs tests against a folder with test classes, not against a jar (and that complicates things)

It does not make sense to me to "rewrite" all tests now. Probably better to apply new approach in areas, which we don't want to back port.